Handle aliases correctly in foreach

mlazos opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/119436

<s>In essence we need to ensure aliases are run in separate foreach kernels so that they are ordered correctly. Previously, aliases could end up in the same kernel which creates weird scheduling dependencies.</s>

There was a bug in cycle detection/can_fuse which was creating cycles when more than two aliases were used in foreach nodes.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

mlazos committed:
Add repro test

mlazos committed:
Add alias tracking to foreach

mlazos committed:
Fix type annotations

mlazos commented on torch/_inductor/lowering.py:
```diff
@@ -448,20 +448,32 @@ def inner_fn(index):
-                        break
+                        if isinstance(t.data, ir.BaseView):
+                            unaliased = t.data.unwrap_view()
+                            alias_counts[unaliased.get_name()] += 1
```
@jansel Am I guaranteed that the underlying storage will have a name? Curious if you have a more canonical approach to get the originating value for a view that's hashable

jansel replied:
No, not everything will have a name.  For example, this could be the result of a random number generation where the RNG generation is fused into the foreach kernel.

jansel reviewed (changes requested):
I think it would be better to deduplicate aliases inside foreach codegen.

Why is aliasing causing a problem?  Normal kernels will deduplicate aliased args automatically.

jansel commented:
Looking at the underlying issue.  I think you need to call `V.kernel.cse.invalidate()` between each kernel.

mlazos commented:
In eager if there's an aliased arg in a foreach arg list, the semantics are the same as if there were a for loop applying the op on pairwise args from left to right. 

So even if I dedup, I need to vertically fuse later uses of the arg into the earlier uses, is that what you're suggesting?

mlazos commented:
> Looking at the underlying issue. I think you need to call `V.kernel.cse.invalidate()` between each kernel.

Won't that just affect vars within the kernel code, not the kernel launcher? I tried calling it between code generating each kernel and it didn't have an effect.

jansel commented:
> Won't that just affect vars within the kernel code, not the kernel launcher? Let me try though

We already have arg dedupping within V.kernel.

mlazos committed:
Merge branch 'main' into mlazos/fix-views

mlazos committed:
Fix pruning of weakdeps for foreach

mlazos commented:
So I figured it out - basically there was a cycle getting introduced because we skipped on checking the dependencies of the whole foreach node on a producer/consumer too early. With aliasing it isn't enough to check that a producer/consumer can be fused into one of the constituent nodes, we need to check that there aren't additional dependencies at the top level that should prevent fusion. I moved the check to later in the scheduler's `can_fuse`. Once I did this, I noticed some optimizer configurations were generating more kernels, I fixed this by pruning Weakdeps at the top level of the `ForeachKernelSchedulerNode`, previously we were only pruning the inner nodes, which was okay, because we only checked fusions at the inner node level.

mlazos committed:
Undo changes to lowering

mlazos committed:
Merge branch 'main' into mlazos/fix-views

mlazos committed:
Merge branch 'main' into mlazos/fix-views

mlazos commented:
@pytorchbot merge
